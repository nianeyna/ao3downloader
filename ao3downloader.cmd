@echo off

pushd %~dp0

if not exist "venv\" (set firstrun="true") else (set firstrun="false")

if %firstrun%=="true" python -m venv venv

call venv\Scripts\activate

if %firstrun%=="true" python -m pip install --upgrade pip & pip install -r requirements.txt

python ao3downloader.py

pause
