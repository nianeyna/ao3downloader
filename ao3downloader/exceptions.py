"""Custom exceptions go here."""


class Ao3DownloaderException(Exception):
    pass


class LockedException(Ao3DownloaderException):
    pass


class DeletedException(Ao3DownloaderException):
    pass


class ProceedException(Ao3DownloaderException):
    pass


class DownloadException(Ao3DownloaderException):
    pass


class LoginException(Ao3DownloaderException):
    pass


class InvalidLinkException(Ao3DownloaderException):
    pass


class InvalidStatusCodeException(Ao3DownloaderException):
    pass
