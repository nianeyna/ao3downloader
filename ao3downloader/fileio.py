"""File operations go here."""

import datetime
import json
import os


# based on https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions
invalid_filename_characters = '<>:"/\|?*.' + ''.join(chr(i) for i in range(32))


def write_log(filename: str, log: dict) -> None:
    log['timestamp'] = datetime.datetime.now().strftime('%m/%d/%Y, %H:%M:%S')
    with open(filename, 'a', encoding='utf-8') as f:
        json.dump(log, f, ensure_ascii=False)
        f.write('\n')


def make_dir(folder: str) -> None:
    if not os.path.exists(folder):
        os.mkdir(folder)


def save_bytes(folder: str, filename: str, content: bytes) -> None:
    file = os.path.join(folder, filename)
    with open(file, 'wb') as f:
        f.write(content)


def get_valid_filename(filename: str) -> str:
    valid_name = filename.translate({ord(i):None for i in invalid_filename_characters})
    return valid_name[:100].strip()


def save_setting(filename: str, setting: str, value) -> None:
    js = get_json(filename)
    if value is None:
        js.pop(setting, None)
    else:
        js[setting] = value
    with open(filename, 'w') as f:
        f.write(json.dumps(js))


def get_setting(filename: str, setting: str):
    js = get_json(filename)
    try:
        return js[setting]
    except:
        return ''


def get_json(filename: str) -> dict:
    with open(filename, 'a', encoding='utf-8'):
        pass
    with open(filename, 'r', encoding='utf-8') as f:
        try:
            return json.load(f)
        except:
            return {}


def setting(prompt: str, filename: str, setting: str):
    value = get_setting(filename, setting)
    if value == '':
        print(prompt)
        value = input()
        save_setting(filename, setting, value)
    return value


def load_logfile(logfile: str) -> list[dict]:
    logs = []
    try:
        with open(logfile, 'r', encoding='utf-8') as f:
            objects = map(lambda x: json.loads(x), f.readlines())
            logs.extend(list(objects))
    except FileNotFoundError:
        pass
    return logs


def file_exists(id: str, titles: dict[str, str], filetypes: list[str], folder: str) -> bool:
    if id not in titles: return False
    filename = get_valid_filename(titles[id])
    files = list(map(lambda x: os.path.join(folder, filename + '.' + x.lower()), filetypes))
    for file in files:
        if not os.path.exists(file):
            return False
    return True
