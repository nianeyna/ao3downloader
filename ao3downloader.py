import ao3downloader.strings as strings

from ao3downloader.actions import ao3download
from ao3downloader.actions import pinboarddownload
from ao3downloader.actions import updatefics
from ao3downloader.actions import redownload
from ao3downloader.actions import logvisualization
from ao3downloader.actions import updateseries
from ao3downloader.actions import getlinks
from ao3downloader.actions import markedforlater


def ao3_download_action():
    ao3download.action()


def links_only_action():
    getlinks.action()


def update_epubs_action():
    updatefics.action()


def update_series_action():
    updateseries.action()
    

def re_download_action():
    redownload.action()


def marked_for_later_action():
    markedforlater.action()


def pinboard_download_action():
    pinboarddownload.action()


def log_visualization_action():
    logvisualization.action()


def display_menu():
    print(strings.PROMPT_OPTIONS)
    for key, value in actions.items():
        try:
            desc = value.description
        except AttributeError:
            desc = value.__name__
        print(' {}: {}'.format(key, desc))


def choose(choice):
    try:
        function = actions[choice]
        try:
            function()
        except Exception as e:
            print(str(e))
    except KeyError as e:
        print(strings.PROMPT_INVALID_ACTION)


display_menu.description = strings.ACTION_DESCRIPTION_DISPLAY_MENU
ao3_download_action.description = strings.ACTION_DESCRIPTION_AO3
update_epubs_action.description = strings.ACTION_DESCRIPTION_UPDATE
pinboard_download_action.description = strings.ACTION_DESCRIPTION_PINBOARD
log_visualization_action.description = strings.ACTION_DESCRIPTION_VISUALIZATION
re_download_action.description = strings.ACTION_DESCRIPTION_REDOWNLOAD
update_series_action.description = strings.ACTION_DESCRIPTION_UPDATE_SERIES
links_only_action.description = strings.ACTION_DESCRIPTION_LINKS_ONLY
marked_for_later_action.description = strings.ACTION_DESCRIPTION_MARKED_FOR_LATER

QUIT_ACTION = 'q'
MENU_ACTION = 'd'

actions = {
    MENU_ACTION: display_menu,
    'a': ao3_download_action,
    'l': links_only_action,
    'u': update_epubs_action,
    's': update_series_action,
    'r': re_download_action,
    'm': marked_for_later_action,
    'p': pinboard_download_action,
    'v': log_visualization_action
    }

display_menu()

while True:
    print('\'{}\' to display the menu again'.format(MENU_ACTION))
    print('please enter your choice, or \'{}\' to quit:'.format(QUIT_ACTION))
    choice = input()
    if choice == QUIT_ACTION: break
    choose(choice)
