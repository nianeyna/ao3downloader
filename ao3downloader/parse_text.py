import datetime

from ao3downloader import strings


def get_pinboard_url(api_token: str, date: datetime.datetime) -> str:
    if date == None:
        return strings.ALL_POSTS_URL.format(api_token)
    else:
        year = str(date.year)
        month = str(date.month).zfill(2)
        day = str(date.day).zfill(2)
        timestamp = strings.TIMESTAMP_URL.format(year, month, day)
        return strings.POSTS_FROM_DATE_URL.format(api_token, timestamp)


def get_valid_filename(filename: str, maximum: int) -> str:
    valid_name = filename.translate({ord(i):None for i in strings.INVALID_FILENAME_CHARACTERS})
    if maximum == 0: return valid_name.strip()
    return valid_name[:maximum].strip()


def get_file_type(filetype: str) -> str:
    return '.' + filetype.lower()


def get_work_number(link: str) -> str:
    return get_digits_after('/works/', link)


def get_series_number(link: str) -> str:
    return get_digits_after('/series/', link)


def is_work(link: str) -> bool:
    return get_work_number(link) != None


def is_series(link: str) -> bool:
    return get_series_number(link) != None


def get_digits_after(test: str, url: str) -> str:
    index = str.find(url, test)
    if index == -1: return None
    digits = get_num_from_link(url, index + len(test))
    if not digits or len(digits) == 0: return None
    return digits


def get_next_page(link: str) -> str:
    index = str.find(link, 'page=')
    if index == -1:
        if str.find(link, '?') == -1:
            newlink = link + '?page=2'
        else:
            newlink = link + '&page=2'
    else:
        i = index + 5
        page = get_num_from_link(link, i)
        nextpage = int(page) + 1
        newlink = link.replace('page=' + page, 'page=' + str(nextpage))
    return newlink


def get_page_number(link: str) -> int:
    index = str.find(link, 'page=')
    if index == -1:
        return 1
    else:
        i = index + 5
        page = get_num_from_link(link, i)
        return int(page)


def get_num_from_link(link: str, start: int) -> str:
    end = start
    while end < len(link) and str.isdigit(link[start:end+1]):
        end = end + 1
    return link[start:end]


def get_total_chapters(text: str, index: int) -> str:
    '''read characters after index until encountering a space.'''
    totalchap = ''
    for c in text[index+1:]:
        if c.isspace():
            break
        else:
            totalchap += c
    return totalchap


def get_current_chapters(text: str, index: int) -> str:
    ''' 
    reverse text before index, then read characters from beginning of reversed text 
    until encountering a space, then un-reverse the value you got. 
    we assume here that the text does not include unicode values.
    this should be safe because ao3 doesn't have localization... I think.
    '''
    currentchap = ''
    for c in reversed(text[:index]):
        if c.isspace():
            break
        else:
            currentchap += c
    currentchap = currentchap[::-1]
    return currentchap


def get_payload(username: str, password: str, token: str) -> dict[str, str]:
    """Get payload for ao3 login."""

    payload = {
        'user[login]': username,
        'user[password]': password,
        'user[remember_me]': '1',
        'authenticity_token': token
    }
    return payload


def get_title_dict(logs: list[dict]) -> dict[str, str]:
    dictionary = {}
    titles = filter(lambda x: 'title' in x and 'link' in x, logs)
    for obj in list(titles):
        link = obj['link']
        if link not in dictionary:
            title = obj['title']
            dictionary[link] = title
    return dictionary


def get_unsuccessful_downloads(logs: list[dict]) -> list[str]:
    links = []
    errors = filter(lambda x:'link' in x and 'success' in x and x['success'] == False, logs)
    for error in errors:
        link = error['link']
        if link not in links: 
            links.append(link)
    return links
