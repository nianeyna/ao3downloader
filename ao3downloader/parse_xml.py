import xml.etree.ElementTree as ET

from urllib.parse import urlparse

from ao3downloader import parse_text


def get_bookmark_list(bookmark_xml: ET.Element, exclude_toread: bool) -> list[dict[str, str]]:
    bookmark_list = []
    for child in bookmark_xml:
        attributes = child.attrib
        # only include valid ao3 links
        link = attributes['href']
        if urlparse(link).hostname == 'archiveofourown.org' and (parse_text.is_work(link) or parse_text.is_series(link)):
            # if exclude_toread is true, only include read bookmarks
            if exclude_toread:
                if not 'toread' in attributes:
                    bookmark_list.append(attributes)          
            # otherwise include all valid bookmarks
            else:
                bookmark_list.append(attributes)
    return bookmark_list


def get_work_link_epub(xml: ET.Element) -> str:
    # assumption: the xml does not contain any links to other works than the one we are interested in. 
    # since this file should not include user-generated html (such as summary) this should be safe.
    # that's a lot of shoulds but we'll let it go because I said so.
    for a in xml.iter('{http://www.w3.org/1999/xhtml}a'):
        href = a.get('href')
        if href and 'archiveofourown.org/works/' in href:
            return href
    return None


def get_stats_epub(xml: ET.Element) -> str:
    # ao3 stores chapter stats in a dd tag with class 'calibre5' for whatever reason.
    for dd in xml.iter('{http://www.w3.org/1999/xhtml}dd'):
        cls = dd.get('class')
        if cls and 'calibre5' in cls:
            return dd.text
    return None


def get_series_epub(xml: ET.Element) -> list[str]:
    series = []
    for a in xml.iter('{http://www.w3.org/1999/xhtml}a'):
        href = a.get('href')
        if href and 'archiveofourown.org/series/' in href:
            series.append(href)
    return series
