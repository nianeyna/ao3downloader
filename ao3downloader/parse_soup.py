import re
import traceback
from typing import Any

from bs4 import BeautifulSoup, ResultSet

from ao3downloader import parse_text, strings
from ao3downloader.exceptions import DownloadException, ProceedException


def get_work_link_html(soup: BeautifulSoup) -> str:
    msg = soup.select('#preface .message a')
    if msg and len(msg) == 2: # there should be exactly two links in here
        return msg[1].get('href') # we want the second one
    return None


def get_stats_html(soup: BeautifulSoup) -> str:
    stats = soup.select('#preface .meta .tags dd')
    for dd in stats:
        if 'Chapters: ' in dd.text:
            return dd.text
    return None


def get_series_html(soup: BeautifulSoup) -> list[str]:
    series = []
    links = soup.select('#preface .meta .tags dd a')
    for link in links:
        href = link.get('href')
        if href and 'archiveofourown.org/series/' in href:
            series.append(href)
    return series


def get_work_link_mobi(soup: BeautifulSoup) -> str:
    # it's ok if there are other work links in the file, because the relevant one will always be the first to appear
    # can't use a more specific selector because the html that comes out of the mobi parser is poorly formatted rip me
    link = soup.find('a', href=lambda x: x and 'archiveofourown.org/works/' in x)
    if link: return link.get('href')
    return None


def get_stats_mobi(soup: BeautifulSoup) -> str:
    stats = soup.find('blockquote', string=lambda x: x and 'Chapters: ' in x)
    if stats: return stats.text
    return None


def get_series_mobi(soup: BeautifulSoup) -> list[str]:
    series = []
    tag = soup.find('p', string=lambda x: x and x == 'Series:')
    if tag:
        block = tag.find_next_sibling('blockquote')
        if block:
            links = block.find_all('a', href=lambda x: x and 'archiveofourown.org/series/' in x)
            for link in links:
                series.append(link.get('href'))
    return series


def get_token(soup: BeautifulSoup) -> str:
    """Get authentication token for logging in to ao3."""

    token = (soup.find('form', class_='new_user')
                 .find('input', attrs={'name': 'authenticity_token'})
                 .get('value'))
    return token


def get_image_links(soup: BeautifulSoup) -> list[str]:
    links = []
    work = soup.find('div', id='workskin')
    if not work: return links
    images = work.find_all('img')
    for img in images:
        href = img.get('src')
        if href:
            links.append(href)
    return links


def get_work_urls(soup: BeautifulSoup) -> list[str]:
    """Get all links to ao3 works on a page"""

    return list(dict.fromkeys(list(
        map(lambda w: get_full_work_url(w.get('href')), 
            filter(lambda a : a.get('href') and parse_text.is_work(a.get('href')), 
                   soup.select('.index.group a'))))))


def get_full_work_url(url: str) -> str:
    """Get full ao3 work url from partial url"""

    work_number = parse_text.get_work_number(url)
    return strings.AO3_BASE_URL + '/works/' + work_number


def get_series_urls(soup: BeautifulSoup, get_all: bool) -> list[str]:
    """Get all links to ao3 series on a page"""

    bookmarks = None if get_all else soup.find_all('li', class_='bookmark')

    return list(dict.fromkeys(list(
        map(lambda w: get_full_series_url(w.get('href')), 
            filter(lambda a : a.get('href') and is_series(a, get_all, bookmarks),
                   soup.select('.index.group a'))))))


def is_series(element: Any, get_all: bool, bookmarks: ResultSet[Any]) -> bool:

    series_number = parse_text.get_series_number(element.get('href'))

    # it's not a series at all, so return false
    if not series_number: return False

    # it is a series and we want all of them, so return true
    if get_all: return True

    # check the bookmarks list to see if this is a series, and return true if it is
    return len(list(filter(lambda x: f'series-{series_number}' in x.get('class'), bookmarks))) > 0


def get_full_series_url(url: str) -> str:
    """Get full ao3 series url from partial url"""

    series_number = parse_text.get_series_number(url)
    return strings.AO3_BASE_URL + '/series/' + series_number


def get_work_and_series_urls(soup: BeautifulSoup, get_all: bool=False) -> list[str]:
    """Get all links to ao3 works or series on a page"""

    work_urls = get_work_urls(soup)
    series_urls = get_series_urls(soup, get_all)
    return work_urls + series_urls


def get_proceed_link(soup: BeautifulSoup) -> str:
    """Get link to proceed through explicit work agreement."""

    try:
        link = (soup.find('div', class_='works-show region')
                    .find('ul', class_='actions')
                    .find('li')
                    .find('a', text=strings.AO3_PROCEED)
                    .get('href'))
    except AttributeError as e:
        raise ProceedException(strings.ERROR_PROCEED_LINK) from e
    return strings.AO3_BASE_URL + link


def get_download_link(soup: BeautifulSoup, download_type: str) -> str:
    """Get download link from ao3 work page."""

    try:
        link = (soup.find('li', class_='download')
                    .find('a', text=download_type)
                    .get('href'))
    except AttributeError as e:
        raise DownloadException(strings.ERROR_DOWNLOAD_LINK) from e
    return strings.AO3_BASE_URL + link


def get_mark_as_read_link(soup: BeautifulSoup) -> str:
    """Get link to mark work as read, if it exists."""

    try:
        link = (soup.find('li', class_='mark')
                    .find('a', text=strings.AO3_MARK_READ)
                    .get('href'))
    except:
        return None
    
    if not link:
        return None

    return strings.AO3_BASE_URL + link


def has_custom_skin(soup: BeautifulSoup) -> bool:
    """Check if a work has custom creator styles"""

    return soup.find('ul', class_='work navigation actions').find('li', class_='style') is not None


def get_title(soup: BeautifulSoup, link: str, pattern: str, pattern_subdir: str) -> str:
    """Get (non-truncated) filename for the work"""

    metadata = get_work_metadata_from_work(soup, link)

    for key, value in metadata.items():
        pattern = pattern.replace(f'{{{key}}}', value)
    pattern = pattern.strip()

    if isinstance(pattern_subdir, str):
        for key, value in metadata.items():
            pattern_subdir = pattern_subdir.replace(f'{{{key}}}', value)
        pattern_subdir = pattern_subdir.strip()

    return (pattern, pattern_subdir)

def get_work_metadata_from_work(soup: BeautifulSoup, link: str) -> dict:
    metadata = {}
    metadata['worknum'] = parse_text.get_work_number(link)
    metadata['title'] = get_text_or_empty(soup, '.preface .title')
    metadata['author'] = get_text_or_empty(soup, '.preface .byline')
    metadata['fandom'] = str.join(', ', list(map(lambda x: x.get_text(), soup.select('dd.fandom a'))))
    metadata['pairing'] = str.join(', ', list(map(lambda x: x.get_text(), soup.select('dd.relationship a'))))
    metadata['rating'] = get_text_or_empty(soup, 'dd.rating')
    metadata['warning'] = str.join(', ', list(map(lambda x: x.get_text(), soup.select('dd.warning a'))))
    metadata['category'] = str.join(', ', list(map(lambda x: x.get_text(), soup.select('dd.category a'))))
    metadata['words'] = get_text_or_empty(soup, 'dd.words').replace(',', '').strip()
    metadata['chapters'] = get_current_chapters(soup)
    metadata['language'] = get_text_or_empty(soup, 'dd.language')
    metadata['published'] = get_text_or_empty(soup, 'dd.published')
    metadata['updated'] = get_text_or_empty(soup, 'dd.status')
    series_list = list(map(lambda x: get_series_from_span(x), soup.select('dd.series span.series span.position')))
    metadata['series_title'] = str.join(', ', list(map(lambda x: x[0], series_list)))
    metadata['series_index'] = str.join(', ', list(map(lambda x: x[1], series_list)))
    metadata['series_number'] = str.join(', ', list(map(lambda x: x[2], series_list)))
    return metadata


def get_text_or_empty(soup: BeautifulSoup, selector: str) -> str:
    """Get text from a selector, or return an empty string if it doesn't exist"""

    try:
        return soup.select(selector)[0].get_text().strip()
    except:
        return ''


def get_series_from_span(soup: BeautifulSoup) -> tuple[str, str, str]:
    """Get series title and index from span element"""

    series_link = soup.find('a')
    series_number = parse_text.get_series_number(series_link['href'])
    series_title = series_link.get_text().strip()
    work_index = re.sub(r'\D', '', soup.decode_contents().replace(str(series_link), '')).strip()
    return series_title, work_index, series_number


def get_work_metadata_from_list(soup: BeautifulSoup, link: str) -> dict:
    metadata = {}
    try:
        worknum = parse_text.get_work_number(link)
        blurb = soup.find('li', class_=f'work-{worknum}')
        tags = blurb.find('ul', class_='tags')
        metadata['title'] = blurb.select('h4.heading a')[0].get_text()
        metadata['author'] = str.join(', ', list(x.get_text() for x in blurb.find_all('a', rel='author')))
        if not metadata['author']: metadata['author'] = 'Anonymous'
        try:
            metadata['summary'] = blurb.find('blockquote', class_='summary').decode_contents()
        except:
            metadata['summary'] = '' # some works don't have a summary
        metadata['fandoms'] = list(x.get_text() for x in blurb.find('h5', class_='fandoms').find_all('a'))
        metadata['warnings'] = list(x.find('a').get_text() for x in tags.find_all('li', class_='warnings'))
        metadata['characters'] = list(x.find('a').get_text() for x in tags.find_all('li', class_='characters'))
        metadata['relationships'] = list(x.find('a').get_text() for x in tags.find_all('li', class_='relationships'))
        metadata['tags'] = list(x.find('a').get_text() for x in tags.find_all('li', class_='freeforms'))
        metadata['words'] = blurb.find('dd', class_='words').get_text()
        metadata['rating'] = blurb.find('span', class_='rating').get_text()
        metadata['chapters'] = blurb.find('dd', class_='chapters').get_text()
        metadata['categories'] = blurb.find('span', class_='category').get_text()
        metadata['complete'] = True if blurb.find('span', class_='iswip').get_text() == 'Complete Work' else False
    except Exception as e: # don't crash the entire download if there is an unhandled exception
        metadata['error'] = ''.join(traceback.TracebackException.from_exception(e).format())
    return metadata


def get_current_chapters(soup: BeautifulSoup) -> str:
    text = (soup.find('dl', class_='stats')
                .find('dd', class_='chapters')
                .get_text().strip())

    index = text.find('/')
    if index == -1: return -1
    
    return parse_text.get_current_chapters(text, index)


def is_locked(soup: BeautifulSoup) -> bool:
    return soup.find('div', id='main', class_='sessions-new') is not None


def is_deleted(soup: BeautifulSoup) -> bool:
    return soup.find('div', id='main', class_='error-404') is not None


def is_explicit(soup: BeautifulSoup) -> bool:
    return soup.find('p', class_='caution') is not None


def is_failed_login(soup: BeautifulSoup) -> bool:
    return string_exists(soup, strings.AO3_FAILED_LOGIN)


def string_exists(soup: BeautifulSoup, string: str) -> bool:
    pattern = string
    expression = re.compile(pattern)
    match = soup.find_all(text=expression)
    return len(match) > 0
