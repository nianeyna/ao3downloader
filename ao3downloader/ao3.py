"""Download works from ao3."""

import os
import traceback

from bs4 import BeautifulSoup

from ao3downloader import exceptions, parse_soup, parse_text, strings
from ao3downloader.fileio import FileOps
from ao3downloader.repo import Repository


class Ao3:
    def __init__(self, repo: Repository, fileops: FileOps, filetypes: list[str], pages: int, series: bool, images: bool, mark: bool=False) -> None:
        self.repo = repo
        self.fileops = fileops
        self.filetypes = filetypes
        self.pages = pages
        self.series = series
        self.images = images
        self.mark = mark
        self.debug = fileops.get_ini_value_boolean(strings.INI_DEBUG_LOGGING, False)


    def download(self, link: str, visited: list[str]=None) -> None:

        log = {}
        if not visited: visited = []

        try:
            self.download_recursive(link, log, visited)
        except Exception as e:
            self.log_error(log, e)


    def update(self, link: str, chapters: str) -> None:
        
        log = {}
        
        try:
            self.download_work(link, log, chapters)
        except Exception as e:
            self.log_error(log, e)


    def update_series(self, link: str, visited: list[str]) -> None:

        log = {}

        try:
            self.download_series(link, log, visited)
        except Exception as e:
            self.log_error(log, e)


    def get_work_links(self, link: str, metadata: bool) -> dict[str, dict]:
        
        links_list = {}
        visited_series = []

        try:
            self.get_work_links_recursive(links_list, link, visited_series, metadata)
        except Exception as e:
            print(strings.ERROR_LINKS_LIST)
            self.log_error({'message': strings.ERROR_LINKS_LIST}, e)
        except KeyboardInterrupt:
            print(strings.INFO_LINKS_LIST_CANCELED)

        return links_list


    def get_work_links_recursive(self, links_list: dict[str, dict], link: str, visited_series: list[str], metadata: bool, soup: BeautifulSoup=None) -> None:

        if parse_text.is_work(link):
            if link not in links_list:
                if metadata:
                    metadata = parse_soup.get_work_metadata_from_list(soup, link)
                    links_list[link] = metadata
                else:
                    links_list[link] = None
        elif parse_text.is_series(link):
            if link not in visited_series:
                visited_series.append(link)
                while True:
                    series_soup = self.repo.get_soup(link)
                    series_soup = self.proceed(series_soup)
                    work_urls = parse_soup.get_work_urls(series_soup)
                    if len(work_urls) == 0: break
                    for work_url in work_urls:
                        self.get_work_links_recursive(links_list, work_url, visited_series, metadata, series_soup)
                    link = parse_text.get_next_page(link)
        elif strings.AO3_BASE_URL in link:
            while True:
                self.fileops.write_log({'link': link, 'message': strings.INFO_STARTING_PAGE, 'level': 'debug'})
                thesoup = self.repo.get_soup(link)
                urls = parse_soup.get_work_and_series_urls(thesoup, self.series)
                if len(urls) == 0:
                    if self.debug: self.fileops.write_log({'link': link, 'message': strings.INFO_NO_WORKS_ON_PAGE, 'level': 'debug'})
                    break
                for url in urls:
                    self.get_work_links_recursive(links_list, url, visited_series, metadata, thesoup)
                link = parse_text.get_next_page(link)
                pagenum = parse_text.get_page_number(link)
                if self.pages and pagenum == self.pages + 1:
                    if self.debug: self.fileops.write_log({'link': link, 'message': strings.INFO_PAGE_LIMIT_REACHED, 'level': 'debug'})
                    break
                print(strings.INFO_FINISHED_PAGE.format(str(pagenum - 1), str(pagenum)))
        else:
            raise exceptions.InvalidLinkException(strings.ERROR_INVALID_LINK)


    def download_recursive(self, link: str, log: dict, visited: list[str]) -> None:

        if link in visited: return
        visited.append(link)

        if parse_text.is_work(link):
            log = {}
            self.download_work(link, log, None)
        elif parse_text.is_series(link):
            log = {}
            self.download_series(link, log, visited)        
        elif strings.AO3_BASE_URL in link:
            while True:
                self.fileops.write_log({'link': link, 'message': strings.INFO_STARTING_PAGE, 'level': 'debug'})
                thesoup = self.repo.get_soup(link)
                urls = parse_soup.get_work_and_series_urls(thesoup, self.series)
                if len(urls) == 0: 
                    if self.debug: self.fileops.write_log({'link': link, 'message': strings.INFO_NO_WORKS_ON_PAGE, 'level': 'debug'})
                    break
                for url in urls:
                    self.download_recursive(url, log, visited)
                if not self.mark:
                    link = parse_text.get_next_page(link)
                    pagenum = parse_text.get_page_number(link)
                    if self.pages and pagenum == self.pages + 1:
                        if self.debug: self.fileops.write_log({'link': link, 'message': strings.INFO_PAGE_LIMIT_REACHED, 'level': 'debug'})
                        break
                    print(strings.INFO_FINISHED_PAGE.format(str(pagenum - 1), str(pagenum)))
        else:
            raise exceptions.InvalidLinkException(strings.ERROR_INVALID_LINK)


    def download_series(self, link: str, log: dict, visited: list[str]) -> None:
        """"Download all works in a series"""

        try:
            while True:
                series_soup = self.repo.get_soup(link)
                series_soup = self.proceed(series_soup)
                work_urls = parse_soup.get_work_urls(series_soup)
                if len(work_urls) == 0: break
                if self.debug: self.fileops.write_log({'link': link, 'message': strings.INFO_STARTING_PAGE, 'level': 'debug'})
                for work_url in work_urls:
                    self.download_recursive(work_url, log, visited)
                link = parse_text.get_next_page(link)
        except Exception as e:
            log['link'] = link
            self.log_error(log, e)


    def download_work(self, link: str, log: dict, chapters: str) -> None:
        """Download a single work"""

        try:
            log['link'] = link
            downloaded = self.try_download(link, log, chapters)
            if downloaded == False: return
        except Exception as e:
            self.log_error(log, e)
        else:
            log['success'] = True
            self.fileops.write_log(log)


    def try_download(self, work_url: str, log: dict, chapters: str) -> bool:
        """Main download logic"""

        thesoup = self.repo.get_soup(work_url)
        thesoup = self.proceed(thesoup)

        if chapters is not None: # TODO this is a super awkward place for this logic to be and I don't like it.
            currentchapters = parse_soup.get_current_chapters(thesoup)
            if int(currentchapters) <= int(chapters):
                return False
        
        pattern = self.fileops.get_ini_value(strings.INI_NAME_PATTERN, strings.INI_DEFAULT_NAME_PATTERN)
        maximum = self.fileops.get_ini_value_integer(strings.INI_NAME_LENGTH, strings.INI_DEFAULT_NAME_LENGTH)
        title = parse_soup.get_title(thesoup, work_url, pattern)
        filename = parse_text.get_valid_filename(title, maximum)
        log['title'] = title
        log['workskin'] = parse_soup.has_custom_skin(thesoup)

        for filetype in self.filetypes:
            link = parse_soup.get_download_link(thesoup, filetype)
            response = self.repo.get_book(link)
            filetype = parse_text.get_file_type(filetype)
            self.fileops.save_bytes(filename + filetype, response)

        if self.images:
            counter = 0
            imagelinks = parse_soup.get_image_links(thesoup)
            for img in imagelinks:
                if str.startswith(img, '/'): break
                try:
                    ext = os.path.splitext(img)[1]
                    if '?' in ext: ext = ext[:ext.index('?')]
                    response = self.repo.get_book(img)
                    imagefile = filename + ' img' + str(counter).zfill(3) + ext
                    self.fileops.save_bytes(os.path.join(strings.IMAGE_FOLDER_NAME, imagefile), response)
                    counter += 1
                except Exception as e:
                    self.fileops.write_log({
                        'message': strings.ERROR_IMAGE, 'link': work_url, 'title': title, 
                        'img': img, 'error': str(e), 'stacktrace': traceback.format_exc()})

        if self.mark:
            marklink = parse_soup.get_mark_as_read_link(thesoup)
            if marklink: self.repo.my_request('GET', marklink)

        return True


    def proceed(self, thesoup: BeautifulSoup) -> BeautifulSoup:
        """Check locked/deleted and proceed through explicit agreement if needed"""

        if parse_soup.is_locked(thesoup):
            raise exceptions.LockedException(strings.ERROR_LOCKED)
        if parse_soup.is_deleted(thesoup):
            raise exceptions.DeletedException(strings.ERROR_DELETED)
        if parse_soup.is_explicit(thesoup):
            proceed_url = parse_soup.get_proceed_link(thesoup)
            thesoup = self.repo.get_soup(proceed_url)
        return thesoup


    def log_error(self, log: dict, exception: Exception):
        log['error'] = str(exception)
        log['success'] = False
        if not isinstance(exception, exceptions.Ao3DownloaderException):
            log['stacktrace'] = ''.join(traceback.TracebackException.from_exception(exception).format())
        self.fileops.write_log(log)
