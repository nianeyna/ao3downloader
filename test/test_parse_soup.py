import os

from bs4 import BeautifulSoup

import ao3downloader.parse_soup as parse_soup


def test_get_work_urls(snapshot):
    soup = get_soup_from_fixture('bookmarks')
    work_urls = parse_soup.get_work_urls(soup)
    assert work_urls == snapshot


def test_get_series_urls_bookmarks(snapshot):
    soup = get_soup_from_fixture('bookmarks')
    series_urls = parse_soup.get_series_urls(soup, False)
    assert series_urls == snapshot


def test_get_series_urls_all(snapshot):
    soup = get_soup_from_fixture('bookmarks')
    series_urls = parse_soup.get_series_urls(soup, True)
    assert series_urls == snapshot


def test_is_locked_true():
    soup = get_soup_from_fixture('lockedWorkLoggedOut')
    assert parse_soup.is_locked(soup) == True


def test_is_locked_false():
    soup = get_soup_from_fixture('lockedWorkLoggedIn')
    assert parse_soup.is_locked(soup) == False


def test_is_deleted_true():
    soup = get_soup_from_fixture('deletedWork')
    assert parse_soup.is_deleted(soup) == True


def test_is_deleted_false():
    soup = get_soup_from_fixture('unlockedWork')
    assert parse_soup.is_deleted(soup) == False


def test_is_explict_true():
    soup = get_soup_from_fixture('explicitWorkLoggedOut')
    assert parse_soup.is_explicit(soup) == True


def test_is_explict_false():
    soup = get_soup_from_fixture('explicitWorkLoggedIn')
    assert parse_soup.is_explicit(soup) == False


def get_soup_from_fixture(filename: str) -> BeautifulSoup:
    fixture_path = os.path.join(os.path.dirname(__file__), 'fixtures', filename + '.html')
    with open(fixture_path) as f:
        soup = BeautifulSoup(f.read(), 'html.parser')
    return soup
