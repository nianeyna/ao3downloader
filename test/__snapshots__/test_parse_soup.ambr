# serializer version: 1
# name: test_get_series_urls_all
  list([
    'https://archiveofourown.org/series/2065602',
    'https://archiveofourown.org/series/3738976',
    'https://archiveofourown.org/series/2627935',
    'https://archiveofourown.org/series/3108957',
    'https://archiveofourown.org/series/3078150',
    'https://archiveofourown.org/series/3078153',
    'https://archiveofourown.org/series/3196530',
    'https://archiveofourown.org/series/2643097',
  ])
# ---
# name: test_get_series_urls_bookmarks
  list([
    'https://archiveofourown.org/series/2065602',
    'https://archiveofourown.org/series/3738976',
  ])
# ---
# name: test_get_work_urls
  list([
    'https://archiveofourown.org/works/34816549',
    'https://archiveofourown.org/works/35778589',
    'https://archiveofourown.org/works/41655369',
    'https://archiveofourown.org/works/34763164',
    'https://archiveofourown.org/works/41214669',
    'https://archiveofourown.org/works/41822007',
    'https://archiveofourown.org/works/342122',
    'https://archiveofourown.org/works/26958667',
    'https://archiveofourown.org/works/33658237',
    'https://archiveofourown.org/works/36398359',
    'https://archiveofourown.org/works/34702543',
    'https://archiveofourown.org/works/35369560',
    'https://archiveofourown.org/works/18623245',
    'https://archiveofourown.org/works/34348333',
    'https://archiveofourown.org/works/28032981',
    'https://archiveofourown.org/works/24412372',
    'https://archiveofourown.org/works/557020',
    'https://archiveofourown.org/works/28968675',
  ])
# ---
